packages/medusa/src/services/draft-order.ts

```diff
@@ -0,0 +1,437 @@
+    shippingOptionService,
+  }: InjectedDependencies) {
+    super()
+
+    this.manager_ = manager
+    this.draftOrderRepository_ = draftOrderRepository
+    this.paymentRepository_ = paymentRepository
+    this.orderRepository_ = orderRepository
+    this.lineItemService_ = lineItemService
+    this.cartService_ = cartService
+    this.productVariantService_ = productVariantService
+    this.shippingOptionService_ = shippingOptionService
+    this.eventBus_ = eventBusService
+  }
+
+  withTransaction(transactionManager: EntityManager): DraftOrderService {
+    if (!transactionManager) {
+      return this
+    }
+
+    const cloned = new DraftOrderService({
+      manager: transactionManager,
+      draftOrderRepository: this.draftOrderRepository_,
+      paymentRepository: this.paymentRepository_,
+      orderRepository: this.orderRepository_,
+      lineItemService: this.lineItemService_,
+      cartService: this.cartService_,
+      productVariantService: this.productVariantService_,
+      shippingOptionService: this.shippingOptionService_,
+      eventBusService: this.eventBus_,
+    })
+
+    cloned.transactionManager_ = transactionManager
+
+    return cloned
+  }
+
+  /**
+   * Retrieves a draft order with the given id.
+   * @param {string} id - id of the draft order to retrieve
+   * @param {object} config - query object for findOne
+   * @return {Promise<DraftOrder>} the draft order
+   */
+  async retrieve(id: string, config = {}): Promise<DraftOrder> {
+    return await this.atomicPhase_(
+      async (transactionManager: EntityManager) => {
+        const draftOrderRepo = transactionManager.getCustomRepository(
+          this.draftOrderRepository_
+        )
+
+        const validatedId = this.validateId_(id)
+
+        const query = this.buildQuery_({ id: validatedId }, config)
+        return await draftOrderRepo.findOne(query).then((draftOrder) => {
+          if (draftOrder) {
+            return draftOrder
+          }
+
+          throw new MedusaError(
+            MedusaError.Types.NOT_FOUND,
+            `Draft order with ${id} was not found`
+          )
+        })
+      }
+    )
+  }
+
+  /**
+   * Retrieves a draft order based on its associated cart id
+   * @param {string} cartId - cart id that the draft orders's cart has
+   * @param {object} config - query object for findOne
+   * @return {Promise<DraftOrder>} the draft order
+   */
+  async retrieveByCartId(cartId: string, config = {}): Promise<DraftOrder> {
+    return await this.atomicPhase_(
+      async (transactionManager: EntityManager) => {
+        const draftOrderRepo = transactionManager.getCustomRepository(
+          this.draftOrderRepository_
+        )
+
+        const query = this.buildQuery_({ cart_id: cartId }, config)
+        return await draftOrderRepo.findOneOrFail(query).catch(() => {
+          throw new MedusaError(
+            MedusaError.Types.NOT_FOUND,
+            `Draft order was not found`
+          )
+        })
```

Comment:
issue: We are introducing yet another way of retrieving from DB here - I would like us to be consistent. See comment on `retrieve` method and refactor.